src/Microsoft.Sbom.Parsers.Spdx30SbomParser/Utils/ElementSerializer.cs

```diff
@@ -0,0 +1,70 @@
+// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System;
+using System.Collections.Generic;
+using System.Text.Json;
+using System.Text.Json.Serialization;
+using System.Xml.Linq;
+using Microsoft.Sbom.Parsers.Spdx30SbomParser.Entities;
+
+public class ElementSerializer : JsonConverter<List<Element>>
+{
+    public override List<Element> Read(ref Utf8JsonReader reader, Type typeToConvert, JsonSerializerOptions options)
+    {
+        throw new NotImplementedException("Deserialization of Elements into specific subtypes is not implemented yet.");
+    }
+
+    public override void Write(Utf8JsonWriter writer, List<Element> elements, JsonSerializerOptions options)
+    {
+        writer.WriteStartArray();
+
+        foreach (var element in elements)
+        {
+            switch (element.Type)
```

Comment:
Haven't tried this, but the internet tells me that it is possible to switch on type - pseudocode:

switch(element)
{
case AnyLicenseInfo:
break;
case CreationInfo:
break;
}

It would make this code less fragile since the case labels could then be compiler verified.

Reply:
Or, should Element expose a Serialize method to be overridden by inheriting classes? This code smells a bit like a problem for polymorphism.

Reply:
+1 for polymorphism here if we can make it work without too much trouble. The current version will break if a class gets renamed and the compiler doesn't know that the class name and the switch statement are somehow linked. The safe non-polymorphism route is to move all of the Type values into a constants class, then use those constants in both the class constructors and in the switch. That's a bit ugly :(

Reply:
No objection to using the as operator instead of casting, but it doesn't really address my primary concern, which is the maintainability of hard-coding the class names as strings.

I notice that JsonSerializer has overloads that allow passing the Type in; would that allow us to write code like this instead?

JsonSerializer.Serialize(writer, element, element.GetType(), options);

That would be a nice elegant solution. Otherwise, but we could implement a virtual method on Element and override it in each derived type; it isn't pretty but with expression-bodied members we can keep it concise and it would eliminate hard-coding as well as keeping type-specific behavior within the type.

Reply:
I ended up using Alison's suggestion - JsonSerializer.Serialize(writer, element, element.GetType(), options); which is much cleaner. Thank you! I'll leave this thread open in case there are any more comments regarding this issue but I think this should solve the previous concerns.